doc/preliminary_report/report.tex

```diff
@@ -0,0 +1,137 @@
+
+% Diviation from proposal
+The investigation into three different algorithms listed in the initial proposal (see attached below) was discarded in favor of a more thorough and deeper look into a single one of them (namely Smith-Waterman). From a computer science perspective all of these algorithms are rather similar, consequently the team decided to focus its efforts on the one which is most relevant in the given domain.
+
+\section{Parallelization strategy}
+In our first attempts two different strategies were pursued that either exploit the inherent structure of the problem or the overall nature of the problem.
+
+\mypar{Fine-grained approach} A major step of the Smith-Waterman algorithm is the calculation of a similarity matrix which measure how closely parts of the tow sequences resemble each other. Each entry in the matrix depends only on its neighbors to the west, north, and north-west. For example, calculating the value encircled in red in Figure 1 only depends on the black encircled entries. Exploiting these data dependencies, each anti-diagonal in the similarity matrix may be calculated independently from each other. This is general approach taken by a number of papers \cite[Sec. 2]{Chen.2010}. We plan to run our benchmark also against those projects as soon as a more representative test example is determined.
+
+The current implementation utilizes OpenMP to split calculations along the anti-diagonal among multiple cores.
+
+\setlength{\extrarowheight}{1.5mm}
+\begin{figure}
+	\centering
+	$\begin{pNiceMatrix}%
+		[name=mymatrix,
+		first-col,
+		first-row,
+		columns-width = 0.3cm
+		]
+		&   & \mathbf{T} & \mathbf{G} & \mathbf{T} & \mathbf{T} \\
+		& 0 & 0 & 0 & 0 & 0 \\
+		\mathbf{G} & 0 & 0 & 3 & 1 & 0 \\
+		\mathbf{G} & 0 & 0 & 3 & 1 & 0 \\
+		\mathbf{T} & 0 & 3 & 1 & 6 & 4 \\
+		\mathbf{T} & 0 & 3 & 1 & 4 & 9 \\
+		
+	\end{pNiceMatrix}$
+	\begin{tikzpicture}[remember picture,overlay]
+		\draw[line width=0.25mm, red](mymatrix-4-4)circle (2mm);
+		\draw[line width=0.25mm](mymatrix-3-4)circle (2mm);
+		\draw[line width=0.25mm](mymatrix-4-3)circle (2mm);
+		\draw[line width=0.25mm](mymatrix-3-3)circle (2mm);
+	\end{tikzpicture}
+	\caption{Data dependencies between entries}
+\end{figure}%
+
+
+
+\mypar{Coarse-grained approach} Usually researchers are not solely interested in how one particular sequence can be aligned optimally to another. Biologically more interesting is the comparison of one  sequence towards a large reference database. For this purpose, we split our mock library file among MPI (message passing interface) nodes and let each one run the algorithm separately. This task is insofar optimal for parallel computers as that no communication between nodes is ever necessary with the exception of generating the output file.

```

Comment:
+ It is also quite often that the reference sequence is much longer than the sample sequence. As we are doing local alignment, the length of the aligned sequence would have an upper bound: not much longer than the sample sequence, so we can break the reference sequence into pieces with small overlaps whose length is the estimation of the upper bound of aligned sequence. Then the original problem can be parallelized by doing the alignment at each pieces. We have implemented this approach with OpenMP which could be a good supplement of the previous MPI approach.

Reply:
This is an approach I have not read in any paper yet. I would rather keep this point for the final report. Would be great if we could present a non-standard approach that actually adds significant speed and which is unmentioned in the scientific literature.

Reply:
But I have already implemented it and it is actually faster.

Reply:
So on leonhard cluster, with 8 pieces of the reference string and 16 threads, sw_solve_small with openmp runs 2x as fast as without.